stdlib/public/core/CString.swift

```diff
@@ -213,9 +258,21 @@ extension String {
     self = string
   }
 
+  /// Creates a new string by copying and validating the null-terminated UTF-8
+  /// data referenced by the given array.
+  ///
+  /// This initializer does not try to repair ill-formed UTF-8 code unit
+  /// sequences. If any are found, the result of the initializer is `nil`.
+  ///
+  /// Note: This initializer is deprecated. Use the initializer
+  ///       `String.init?(validating: array, as: UTF8.self)` instead,
+  ///       remembering that "\0" is a valid character in Swift.
```

Comment:
The fact that the trailing null is going to get absorbed into the Swift string makes this _not_ a drop-in replacement. Developers will see the deprecation message, but not the note about null characters. I would suggest instead telling developers to use `array.withUnsafeBuffer {}` so they can use the other `validatingCString:` initializer that takes a pointer.

Reply:
If there is a known trailing zero, one can use `String(validating: myTerminatedArray.dropLast(), as: UTF8.self)`, and that's better than using the unsafe hatch. I don't think the note should be overly prescriptive, though.

Reply:
I _can_ do that, sure, but if I'm unfamiliar with the differences in API, I'm presumably just going to do what the diagnostic tells me to (replace this call with one to the other API) and then I'll introduce a correctness bug that I might not even notice at first.

We don't need an overly verbose diagnostic, but the one we emit should provide clear and correct instructions for resolving the issue. :)